TrkDiag/src/TrackMatching_module.cc

```diff
@@ -0,0 +1,304 @@
+              }
+
+              // add the tracks to the clusters
+              add_tracks(clusters, cluster_index, itrkptr, jtrkptr);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  void TrackMatching::produce(art::Event& event ) {
+    // Create the output
+    std::unique_ptr<KalSeedPtrCollections> ptr_clusters(new KalSeedPtrCollections);
+
+    // Get the track collections
+    std::vector<const KalSeedCollection*> trackColls;
+    std::vector<art::ValidHandle<KalSeedCollection>> handles;
+    for(auto tag : _seedTags) {
+      auto handle = event.getValidHandle<KalSeedCollection>(tag);
+      handles.push_back(handle);
+      trackColls.push_back(handle.product());
+      if(_debug > 0) printf("[TrackMatching::%s::%s] Track collection %s has %zu entries\n",
+                            __func__, moduleDescription().moduleLabel().c_str(),
+                            tag.encode().c_str(), handle->size());
+    }
+
+    //
+    // For each track in each collection, check for overlaps with each other track
+    //
+
+    std::map<const KalSeed*, KalSeedPtr> ptr_map; // map of track -> art::Ptr
+    std::vector<std::vector<const KalSeed*>> clusters; // list of clusters of matched tracks
+    std::map<const KalSeed*, size_t> cluster_index; // map of track -> index in the cluster list
+
+    create_clusters(trackColls, handles, ptr_map, clusters, cluster_index);
+
+    //
+    // From the clustered track lists, create the output track cluster collections
+    //
+
+    // Create the matched clusters
+    int nclusters = 0; // count the non-empty clusters
+    for(auto& cluster : clusters) {
+      if(cluster.empty()) continue; // skip emptied clusters due to cluster merging
+      if(_debug > 0) printf("  Adding a track cluster with %zu tracks\n", cluster.size());
+      if(_makeHists) _hists.cluster_sizes->Fill(cluster.size());
+      ++nclusters;
+      KalSeedPtrCollection ptrs;
+      for(auto trk : cluster) ptrs.push_back(ptr_map[trk]);
+      ptr_clusters->push_back(ptrs);
+    }
+
+
+    // Add non-overlapping tracks as single track clusters so all tracks are in the output
+    int nsingle_clusters = 0; // count the isolated tracks
+    for(auto entry : ptr_map) {
+      if(cluster_index.count(entry.first)) continue;
+      if(_debug > 0) printf("  Adding a single track cluster\n");
+      ++nsingle_clusters;
+      KalSeedPtrCollection ptrs;
+      ptrs.push_back(entry.second);
+      if(_makeHists) _hists.cluster_sizes->Fill(1);
+      ptr_clusters->push_back(ptrs);
+    }
+
+    if(_debug > 0) printf("[TrackMatching::%s::%s] Found %i track clusters and %i single tracks\n",
+                          __func__, moduleDescription().moduleLabel().c_str(), nclusters, nsingle_clusters);
+    if(_makeHists) _hists.nclusters->Fill(nclusters + nsingle_clusters);
+
```

Comment:
I suggest adding a test that every input track ends up in exactly 1 cluster.

I believe you could simplify the flow by starting by putting every input track into its own cluster. Then the clustering would proceed purely as merges, using a double-nested pairing loop ,instead of quadruple-nesting. This has the advantage of working directly with the output product, and should allow you to remove some (all?) of the intermediate bookkeeping structures.

Reply:
This is definitely a cleaner logic, I've cleaned up a lot of the code by making this simplification. I've also added a test that N(in tracks) == N(out tracks), but I haven't yet added a test that every input KalSeedPtr is in the end clustered. Should I add this test as well?

Reply:
Your test should be adequate.